Added Services through dependency injection

ZachCarrillo opened the pull request:
Added Services to my project, Entity Framework was added a previous week so no change there.

ZachCarrillo committed:
Created ContactBookContext

Combined all previous Context Classes into ContactBookContext

ZachCarrillo committed:
updated Post method for Groupmember controller

made post method check if the ids of customer and group exist before adding to database

ZachCarrillo committed:
Added CustomersServices

Added Services for customer model and updated customer controller to use service

ZachCarrillo committed:
Added GroupService and updated GroupMembersController

ZachCarrillo committed:
Add Members Service and update GroupMembers controller

ysemerikov commented on CustomerContactBook/Services/GroupService.cs:
```diff
@@ -0,0 +1,71 @@
+            _context = context;
+        }
+
+        public async Task<ActionResult<IEnumerable<CustomerGroup>>> GetGroups()
+        {
+            return await _context.Groups.ToListAsync();
```
Is it compilable? I'm not sure, because the method has signature `ActionResult<IEnumerable<CustomerGroup>>` but you returns `List<CustomerGroup>`.
So there is couple of important points:
1. `ActionResult` is only for controller level, services shouldn't work with that.
2. If service returns `CustomerGroup` (database model) it means controller operates with database models, that's bad practice. The controller should work with another models. In our case it may looks weird - you will just copy all fields of one class to another, but it gives a lot of good points:
* external caller doesn't have access to database model = it can't change data;
* it's easier to find usages / do refactoring;
* when you add new field in database it will not automatically appear in the API;
* vise versa - to add new API field you don't need to change database.

In the enterprise applications 3 model lawyers are used: database classes, service classes, controller classes. So service manipulates database classes, but always returns another class (service model). Controller get service models, but returns another class (controller model). It allows to reuse one service in the different controllers/places and returns different models. Also it's easier for refactoring.
In our project I don't want to bring all 3 lawyers, I think 2 are enough - first is controller+service and the second is database.

ysemerikov replied:
Please fix at least the first point, then I will continue to review.

ZachCarrillo replied:
I made it so services no longer return ActionResults or Database models in my latest commit

ysemerikov replied:
Good job!

ZachCarrillo committed:
Updated Services and Seperated Service and database models

Updated services to not return ActionResult.
Seperated models into database and service models. service classes return service models to controllers

ysemerikov commented on CustomerContactBook/Controllers/CustomerGroupsController.cs:
```diff
@@ -57,25 +55,16 @@ public async Task<ActionResult<CustomerGroup>> GetCustomerGroup(int id)
-        public async Task<IActionResult> PutCustomerGroup(int id, CustomerGroup customerGroup)
+        public async Task<IActionResult> PutCustomerGroup(int id, CustomerGroupModel customerGroup)
         {
             if (id != customerGroup.Id)
```
I don't recommend having `id` in the update model - this parameter should be only in the query path.

ysemerikov commented on CustomerContactBook/Controllers/CustomerGroupsController.cs:
```diff
@@ -85,12 +74,11 @@ public async Task<IActionResult> PutCustomerGroup(int id, CustomerGroup customer
+            var result = await _groupService.PostCustomerGroup(customerGroup);
 
-            return CreatedAtAction("GetCustomerGroup", new { id = customerGroup.Id }, customerGroup);
+            return CreatedAtAction("GetCustomerGroup", new { id = result.Id }, result);
```
I like it!

ysemerikov commented on CustomerContactBook/Services/CustomersService.cs:
```diff
@@ -0,0 +1,122 @@
+            {
+                return null;
+            }
+            _context.Entry(customer).State = EntityState.Modified;
```
it's really tricky thing! I like it!
But please try to avoid complex things in the real projects - usually the cost of maintaining the code is higher than the amount of resources you save (CPU/network). That is, the other person will spend more time trying to understand what is written here. And even worse - they will not understand and create a bug.

ysemerikov commented on CustomerContactBook/Services/CustomersService.cs:
```diff
@@ -0,0 +1,122 @@
+            return toCustomerModel(customer);
+        }
+
+        public async Task<bool?> PutCustomer(long id, CustomerModel model)
```
I don't see you use `false`, maybe you don't need `null`?

ysemerikov commented on CustomerContactBook/Services/CustomersService.cs:
```diff
@@ -0,0 +1,122 @@
+        {
+            var customer = new Customer
+            {
+                Id = model.Id,
```
This is a fully working way to set an id, but it's always better to let DB takes care of it.

ysemerikov commented on CustomerContactBook/Services/CustomersService.cs:
```diff
@@ -0,0 +1,122 @@
+            return _context.Customers.Any(e => e.Id == id);
+        }
+
+        private static CustomerModel toCustomerModel(Customer customer)
```
in C# all methods start with capital letter.

ysemerikov commented on CustomerContactBook/Services/GroupService.cs:
```diff
@@ -0,0 +1,107 @@
+            _context = context;
+        }
+
+        public async Task<List<CustomerGroupModel>> GetGroups()
```
in the other methods you use `CustomerGroup`, maybe add `Customer` word here as well?

ysemerikov commented on CustomerContactBook/Services/GroupService.cs:
```diff
@@ -0,0 +1,107 @@
+            return toGroupModel(customerGroup);
+        }
+
+        public async Task<bool?> PutCustomerGroup(long id, CustomerGroupModel model)
```
Put is HTTP method and it's clear why you may use it in the controller's method names. But there is service layer and better to use regular language (update, create, ...)

ZachCarrillo committed:
Updated variable names, method names, and changed put method return values

updated variable names, made methods start with a capital letter, and changed bool? to bool and returned false instead of null in put methods
